ambry-router/src/main/java/com.github.ambry.router/CryptoWorker.java

```diff
@@ -0,0 +1,296 @@
+ */
+package com.github.ambry.router;
+
+import com.github.ambry.commons.BlobId;
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Thread responsible for encrypting and decrypting blob content and per blob keys. Worker Thread listens to a job queue
+ * for new jobs and processes it one by one.
+ */
+public class CryptoWorker implements Runnable {
+  private final BlockingQueue<CryptoJob> jobQueue;
+  private final CryptoService cryptoService;
+  private final KeyManagementService kms;
+  private final AtomicBoolean enabled;
+  private CountDownLatch shutdownLatch;
+
+  private static final Logger logger = LoggerFactory.getLogger(CryptoWorker.class);
+
+  /**
+   * Instantiates {@link CryptoWorker} with the job queue, {@link CryptoService} and {@link KeyManagementService}
+   * @param jobQueue the {@link BlockingQueue} that contains the jobs
+   * @param cryptoService the {@link CryptoService} to use to encrypt or decrypt
+   * @param kms the {@link KeyManagementService} to fetch keys from
+   */
+  CryptoWorker(BlockingQueue<CryptoJob> jobQueue, CryptoService cryptoService, KeyManagementService kms) {
+    this.jobQueue = jobQueue;
+    this.cryptoService = cryptoService;
+    this.kms = kms;
+    enabled = new AtomicBoolean(false);
+  }
+
+  @Override
+  public void run() {
+    enabled.set(true);
+    shutdownLatch = new CountDownLatch(1);
+    try {
+      while (enabled.get()) {
+        try {
+          CryptoJob cryptoJob = jobQueue.poll(10, TimeUnit.MILLISECONDS);
+          if (cryptoJob != null) {
+            cryptoJob.doOperation(cryptoService, kms);
+          }
+        } catch (InterruptedException e) {
+          logger.error("InterruptedException thrown in CryptoWorker");
+        }
+      }
+    } finally {
+      shutdownLatch.countDown();
+    }
+  }
+
+  /**
+   * Shuts down the thread by closing all pending jobs
+   */
+  void close() {
+    if (enabled.compareAndSet(true, false)) {
+      logger.info("Closing CryptoWorker");
+      closePendingJobs(new GeneralSecurityException("Shutting down EncryptDecrypt Thread"));
+    }
+    if (shutdownLatch != null) {
+      try {
+        shutdownLatch.await(1, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        logger.error("InterruptedException thrown while waiting for CryptoWorker to shutdown");
+      }
+    }
+  }
+
+  /**
+   * Closes all pending jobs with the given {@code e}
+   * @param gse the {@link GeneralSecurityException} that needs to be set while invoking callbacks for the pending jobs
+   */
+  private void closePendingJobs(GeneralSecurityException gse) {
+    CryptoJob cryptoJob = jobQueue.poll();
+    while (cryptoJob != null) {
```

Comment:
Is this bounded when this method is called? (Basically, can jobs keep getting added to this queue?) If not, the "poison" idiom works well here too.

Reply:
Plan is to have a JobHandler which is the interface to the callers. I will sync up offline on this. But the gist is that, the JobHandler will ensure no new jobs will be added to the queue after shutdown.